### PR TITLE
[WIP] - Build vpc

### PR DESCRIPTION
**Pull request by victortrac**
Creating a PR for review comments

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,12 @@
+.idea
+*.pyc
+
+.git-crypt.key
```
Comment (victortrac) [inline .gitignore]: This is not used. Please remove.

File: terraform/README.md [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+```
+$ terraform
+Usage: terraform [--version] [--help] <command> [args]
+
+The available commands for execution are listed below.
+...
+
+Common commands:
+    apply              Builds or changes infrastructure
+    console            Interactive console for Terraform interpolations
```
Comment (victortrac) [inline terraform/README.md]: plan?

File: terraform/deployment_services.tf.dev [hunk 3]
```diff
@@ -0,0 +1,7 @@
+data "terraform_remote_state" "prod" {
+  backend = "gcs"
+  config = {
+    bucket = "discovery-etcd-io-prod-infrastructure"
```
Comment (victortrac) [inline terraform/deployment_services.tf.dev]: I prefer pattern: `<project>-infrastructure`. It is easy to remember and simpler.

File: terraform/environments/prod/config.tfvars [hunk 4]
```diff
@@ -0,0 +1,7 @@
+environment = "prod"
+
+gcp = {
+  project = "etcd-io-prod"
```
Comment (victortrac) [inline terraform/environments/prod/config.tfvars]: The prod project is `etcd-io`.

File: terraform/main.tf [hunk 5]
```diff
@@ -0,0 +1,10 @@
+provider "google" {
+  version = "~> 2.10.0"
```
Comment (victortrac) [inline terraform/main.tf]: Please verify the latest version of this provider.

File: terraform/remote_data.tf.prod [hunk 6]
```diff
@@ -0,0 +1,7 @@
+data "terraform_remote_state" "dev" {
+  backend = "gcs"
+  config = {
+    bucket = "discovery-etcd-io-dev-infrastructure"
```
Comment (victortrac) [inline terraform/remote_data.tf.prod]: I prefer pattern: `<project>-infrastructure`. It is easy to remember and simpler.

**Review by victortrac — COMMENTED** [review]
@eduartua Please see comments.

**Comment by eduartua** [comment]
All suggested changes were applied.

**Comment by victortrac** [comment]
I also forgot to mention that none of the environment specific files (like deployment_services.tf.dev) will apply, because those are dynamically symlinked by our terrafor.sh. If you're not planning on using those files, then you need to figure out another solution.

**Comment by eduartua** [comment]
> I also forgot to mention that none of the environment specific files (like deployment_services.tf.dev) will apply, because those are dynamically symlinked by our terrafor.sh. If you're not planning on using those files, then you need to figure out another solution.

This is the solution I have come up:

- I have declared two new variables `reference_prod_data` and `reference_dev_data` of type bool, and the default is set to _false_.
- We are going to assign _true_ according to the environment and the remote we want to interact with. e.g. In `dev` we want to get data from `prod`, so that we set `reference_prod_data=true` in `environment/dev/config.tfvars`. [see](https://github.com/cloudkite-io/discovery.etcd.io/blob/build-vpc/terraform/environments/dev/config.tfvars#L3)
- Then, when creating resources with data from `prod` we can use the `count` in a conditional expression like this:
```
# Support gcr.io
resource "google_storage_bucket_iam_binding" "gcr-io" {
  count = var.reference_prod_data ? 1 : 0

  bucket = "artifacts.${var.gcp["project"]}.appspot.com"
  role = "roles/storage.objectViewer"
  members = [
        "serviceAccount:${google_service_account.gke_service_account.email}",
        "serviceAccount:${data.terraform_remote_state.prod[count.index].outputs.gke_service_account_email}",
  ]
}
```

If `var.reference_prod_data` is `true`, the count parameter for the `google_storage_bucket_iam_binding` resource will be set to 1, so it will be created. If `var.reference_prod_data` is false, the count parameter the `google_storage_bucket_iam_binding` resource will be set to 0, so it will not be created.

I tested this solution and seems to work fine.

**Comment by victortrac** [comment]
This is pretty ugly. Let me think about what we can do.

**Comment by eduartua** [comment]
I agree. That was just the solution I found today. I'll try to do it in another way.

**Comment by eduartua** [comment]
Now it is updated to use the file layout per environment. This has been tested in dev and the plan output looks good.

**Pull request merged**